#Conceitos iniciais de Python


##Operações

1. Aritméticas

In [1]:
soma = 1 + 3
print(soma)

4


In [2]:
a = 2
b = 5

soma = a + b

print (soma)

7


In [3]:
produto = 2*3

razao = 6/3

pow = b**a

print("Este é o produto de 2 e 3: {}, esta a razão entre 6 e 3: {}, e esta é o resultado de 5 ao quadrado: {}".format(produto, razao, pow))


Este é o produto de 2 e 3: 6, esta a razão entre 6 e 3: 2.0, e esta é o resultado de 5 ao quadrado: 25


2. Booleanas

In [4]:
print(4 < 5)
print(4 > 5)

True
False


In [5]:
#É possível comparar com variáveis também!

a = 10
b = 12

print("a > b?", a > b)

print ("5 é igual a 5?", 5 == 5)

#Podemos armazenar o resultado de alguma comparação em uma outra variável para utilizá-la futuramente (importante para loops, por exemplo)

c = a == b #a variável c armazena o resultado da comparação 'a é igual a b?'
d = b != 10 #d recebe o resultado da comparação "b é diferente de 10?"
print("a é igual a b?", c)
print("b é diferente de 10?", d)

a > b? False
5 é igual a 5? True
a é igual a b? False
b é diferente de 10? True


##Loops e condicionais

1. Loops

In [6]:
#Calculando um produto de um número por ele mesmo utilizando um loop while

nro = 4
rep = 4

while rep != 0:
  resultado = 4*4
  rep = rep - 1
  print("Faltam {} repetições".format(rep))
else:
  print("resultado:", resultado)


#Como seria com um loop for?
resultado2 = 0
for i in range(1, 5):
  resultado2 += 4
  print("repetição i =", i, "resultado = ", resultado2)

Faltam 3 repetições
Faltam 2 repetições
Faltam 1 repetições
Faltam 0 repetições
resultado: 16
repetição i = 1 resultado =  4
repetição i = 2 resultado =  8
repetição i = 3 resultado =  12
repetição i = 4 resultado =  16


In [7]:
#Como o for interage com uma lista por exemplo?

comidas = []

comidas.append("banana")
comidas.append("maça")
comidas.append('uva')

print(comidas)

for i in comidas: #itera sobre cada item da lista. Útil para futuras operações com os itens
  print(i)

['banana', 'maça', 'uva']
banana
maça
uva


In [8]:
#E uma lista de palavras iterando com um for?

entradas = []

for i in range(0, 5): #um for que irá repetir 4 vezes
  entradas.append(input())


 1
 2
 3
 4
 5


In [9]:
print(entradas[0], "---", type(entradas[0]))
print(entradas[2], "---", type(entradas[2]))
print(entradas[4], "---", type(entradas[4]))

#Pera, 1 não é pra ser um inteiro?

#entradas[0] = int(entradas[0]) #Convertendo o valor que está contido na lista entradas, na posição 0, para o tipo inteiro

#print(entradas[0], "---", type(entradas[0]))

1 --- <class 'str'>
3 --- <class 'str'>
5 --- <class 'str'>


2. Condicionais

In [10]:
valor = 3

if valor > 10: #Verifica se a variável valor é maior que 10
  print("Valor é maior que 10!") #se for, realiza uma ação
else: #se não for, realiza outra
  print("Valor é menor que 10 :(")

#É possível definir mais de um condicional por exemplo:

valor1 = 23
valor2 = 10

if valor1 == 23 and valor2 == 10:
  print("Tá correto!!")
elif valor1 == 23 and valor2 != 10:
  print("Valor 1 ok")
elif valor1 != 23 and valor2 == 10:
  print("valor 2 ok")
else:
  print("Tudo incorreto!")


Valor é menor que 10 :(
Tá correto!!


##Dicionários

In [11]:
#Exemplo de dicionários:

identificaçao = {'daniel':1, 'mariana':2, 'cesar':3, 'renata':4, 'lachi':4} #Os nomes são as keys e os números os valores

print(identificaçao['cesar']) #podemos acessar cada os valores por meios das keys a eles associadas


#É possível associar mais de um valor a uma mesma key
membros  = {'estrategia':['daniel', 'cesar', 'lachi', 'mariana', 'sasaki'], 'nlp':['renata', 'felipe', 'lachi', 'Rian'], 'ds':['kauã', 'daniel', 'davi', 'feitosa']}

print(membros['estrategia'])
print(membros['estrategia'][0]) #E ainda acessar a cada uma delas individualmente


3
['daniel', 'cesar', 'lachi', 'mariana', 'sasaki']
daniel


##Funções


In [12]:
#Basicamente, há dois tipos de funções: as que retornam algum valor e aquelas que não retornam

def calculaProduto(nro1, nro2): #Transformando em função aquele mesmo laço while do início
  resultado = 0
  while nro2 != 0:
    resultado += nro1
    nro2 -= 1

  return resultado

produto2e3 = calculaProduto(2, 3)

print(produto2e3)


def separaLista (listaOriginal, listaPar, listaImpar): #Recebe duas listas e separa os valores conforme o valor em que cada um está registrado
  for i in range(0, len(listaOriginal) - 1, 2):
    listaPar.append(listaOriginal[i])
  for i in range(1, len(listaOriginal) - 1, 2):
    listaImpar.append(listaOriginal[i])

listaOriginal = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
lista1 = []
lista2 = []

separaLista(listaOriginal, lista1, lista2)

print(listaOriginal, '\n', lista1, '\n', lista2)

6
[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15] 
 [0, 2, 4, 6, 8, 10, 12, 14] 
 [1, 3, 5, 7, 9, 11, 13]


# Pandas

- A ideia deste notebook é fazer uma introdução ao [Pandas](https://pandas.pydata.org/), uma biblioteca que será muito usada no contexto da análise de dados.

In [13]:
import pandas as pd #importando o pandas

A análise começa sempre importando um arquivo. O Pandas aceita vários formatos diferentes, mas o mais comum costuma ser csv.
Então, instanciamos uma variável, aqui chamada de "data", que receberá as informações do dataset lido pelo pandas

In [14]:
data = pd.read_csv("./adult.csv", na_values='?')

Após carregar o dataset, vamos dar uma olhada inicial nas informações contidas nele.

Com a função `.info()` é possível entender como o dataset está organizado, quantas linhas e colunas (features) há, bem como o tipo dos dados registrados em cada uma delas. Interessante, por exemplo, por permitir entender se é necessário alterar o tipo de alguma feature.

Já as funções `.head()` e `.tail()` permitem uma observação de algumas linhas iniciais e finais, respectivamente, do dataset. Costuma ser bem útil para entender um pouco mais da disposição dos dados, bem como para ter ideias de como trabalhar com as informações disponíveis. A quantidade de linhas mostradas na sequência de cada um dos comandos, por padrão, é 10, mas o valor pode ser alterado passando um número.

In [15]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48842 entries, 0 to 48841
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   age              48842 non-null  int64 
 1   workclass        46043 non-null  object
 2   fnlwgt           48842 non-null  int64 
 3   education        48842 non-null  object
 4   educational-num  48842 non-null  int64 
 5   marital-status   48842 non-null  object
 6   occupation       46033 non-null  object
 7   relationship     48842 non-null  object
 8   race             48842 non-null  object
 9   gender           48842 non-null  object
 10  capital-gain     48842 non-null  int64 
 11  capital-loss     48842 non-null  int64 
 12  hours-per-week   48842 non-null  int64 
 13  native-country   47985 non-null  object
 14  income           48842 non-null  object
dtypes: int64(6), object(9)
memory usage: 5.6+ MB


In [16]:
data.shape

(48842, 15)

Como resultado da função .info(), nota-se que não há nenhum dado faltante em todas as 48842 linhas do dataset. Além disso, percebe-se que há mais features não-numéricas do que numéricas. Vamos observar mais de perto a organização dessas features

In [17]:
data.head()

age  workclass  fnlwgt     education  educational-num      marital-status  \
0   25    Private  226802          11th                7       Never-married   
1   38    Private   89814       HS-grad                9  Married-civ-spouse   
2   28  Local-gov  336951    Assoc-acdm               12  Married-civ-spouse   
3   44    Private  160323  Some-college               10  Married-civ-spouse   
4   18        NaN  103497  Some-college               10       Never-married   

          occupation relationship   race  gender  capital-gain  capital-loss  \
0  Machine-op-inspct    Own-child  Black    Male             0             0   
1    Farming-fishing      Husband  White    Male             0             0   
2    Protective-serv      Husband  White    Male             0             0   
3  Machine-op-inspct      Husband  Black    Male          7688             0   
4                NaN    Own-child  White  Female             0             0   

   hours-per-week native-country income  
0              40  United-States  <=50K  
1              50  United-States  <=50K  
2              40  United-States   >50K  
3              40  United-States   >50K  
4              30  United-States  <=50K

In [18]:
data.tail()

age     workclass  fnlwgt   education  educational-num  \
48837   27       Private  257302  Assoc-acdm               12   
48838   40       Private  154374     HS-grad                9   
48839   58       Private  151910     HS-grad                9   
48840   22       Private  201490     HS-grad                9   
48841   52  Self-emp-inc  287927     HS-grad                9   

           marital-status         occupation relationship   race  gender  \
48837  Married-civ-spouse       Tech-support         Wife  White  Female   
48838  Married-civ-spouse  Machine-op-inspct      Husband  White    Male   
48839             Widowed       Adm-clerical    Unmarried  White  Female   
48840       Never-married       Adm-clerical    Own-child  White    Male   
48841  Married-civ-spouse    Exec-managerial         Wife  White  Female   

       capital-gain  capital-loss  hours-per-week native-country income  
48837             0             0              38  United-States  <=50K  
48838             0             0              40  United-States   >50K  
48839             0             0              40  United-States  <=50K  
48840             0             0              20  United-States  <=50K  
48841         15024             0              40  United-States   >50K

De cara, é possível perceber que, apesar de não estar registrada a ocorrência de dados faltantes, há, na linha 4, um ponto de interrogação como resposta para a classe do trabalhador em questão. Além disso, vê-se também uma certa imprecisão em algumas respostas da feature *education*, algumas respostas estão num padrão e outras em outro.
Por fim, com essas saídas do programa, já temos conhecimento suficiente para afirmar como estão organizadas as informações e quais delas seriam interessantes manter ou retirar.

###Series
O comando Series do Pandas é extremamente importante pois permite uma visualização de todo o dataset de maneira mais concisa em um vetor de uma única dimensão, facilitando o trabalho de edição e, quando necessário, o de observação do próprio dataset.

Com o comando `.head()`, vimos que há pelo menos um problema com a feature *worcklass*. Vamos analisar essa feature um pouco mais de perto para entender quais são todas as informações inseridas nesse tópico.

Com o comando `.unique()`, conseguimos uma listagem simples, em uma única dimensão, de todas as respostas que estão presentes no dataset.

In [19]:
pd.Series(data['workclass'].unique())

0             Private
1           Local-gov
2                 NaN
3    Self-emp-not-inc
4         Federal-gov
5           State-gov
6        Self-emp-inc
7         Without-pay
8        Never-worked
dtype: object

In [20]:
pd.Series(data.education.unique())

0             11th
1          HS-grad
2       Assoc-acdm
3     Some-college
4             10th
5      Prof-school
6          7th-8th
7        Bachelors
8          Masters
9        Doctorate
10         5th-6th
11       Assoc-voc
12             9th
13            12th
14         1st-4th
15       Preschool
dtype: object

In [21]:
pd.Series(data["marital-status"].unique())

0            Never-married
1       Married-civ-spouse
2                  Widowed
3                 Divorced
4                Separated
5    Married-spouse-absent
6        Married-AF-spouse
dtype: object

In [22]:
pd.Series(data['native-country'].unique())

0                  United-States
1                            NaN
2                           Peru
3                      Guatemala
4                         Mexico
5             Dominican-Republic
6                        Ireland
7                        Germany
8                    Philippines
9                       Thailand
10                         Haiti
11                   El-Salvador
12                   Puerto-Rico
13                       Vietnam
14                         South
15                      Columbia
16                         Japan
17                         India
18                      Cambodia
19                        Poland
20                          Laos
21                       England
22                          Cuba
23                        Taiwan
24                         Italy
25                        Canada
26                      Portugal
27                         China
28                     Nicaragua
29                      Honduras
30        

In [23]:
data['native-country'].isna().sum()

857

In [24]:
data['native-country'].mode()

0    United-States
Name: native-country, dtype: object

In [25]:
data['workclass'].isna().sum()

2799

In [26]:
data['workclass'].dropna(inplace = True)

In [27]:
data.loc[data['native-country'].isna(), 'education'].value_counts()

education
Bachelors       210
HS-grad         198
Some-college    155
Masters          78
Doctorate        32
Prof-school      26
7th-8th          24
Assoc-voc        23
Assoc-acdm       22
12th             22
5th-6th          19
10th             17
11th             12
9th              11
1st-4th           7
Preschool         1
Name: count, dtype: int64

In [28]:
data.loc[(data['native-country'] == 'United-States') & (data['income'] == '>50K'), 'gender'].value_counts()

gender
Male      9081
Female    1613
Name: count, dtype: int64

In [29]:
data[(data['age'] >= 80) | (data['age'] < 20)].head()

age workclass  fnlwgt     education  educational-num marital-status  \
4    18       NaN  103497  Some-college               10  Never-married   
38   17   Private  269430          10th                6  Never-married   
50   18   Private   54440  Some-college               10  Never-married   
67   19   Private  105460  Some-college               10  Never-married   
75   17       NaN  165361          10th                6  Never-married   

           occupation   relationship   race  gender  capital-gain  \
4                 NaN      Own-child  White  Female             0   
38  Machine-op-inspct  Not-in-family  White    Male             0   
50      Other-service      Own-child  White    Male             0   
67      Other-service      Own-child  White    Male             0   
75                NaN      Own-child  White    Male             0   

    capital-loss  hours-per-week native-country income  
4              0              30  United-States  <=50K  
38             0              40  United-States  <=50K  
50             0              20  United-States  <=50K  
67             0              20  United-States  <=50K  
75             0              40  United-States  <=50K

In [30]:
print("Pessoas com etnia diferente de 'Black' e 'White' que têm renda maior que 50K")

data_cp = data.copy()
data_cp['is_from_us'] = data_cp['native-country'] == 'United-States'
data_cp.loc[(~data_cp['race'].isin(['Black', 'White'])) & (data['income'] == '>50K'), ['gender', 'is_from_us']].value_counts()

Pessoas com etnia diferente de 'Black' e 'White' que têm renda maior que 50K


gender  is_from_us
Male    False         274
        True          145
Female  False          54
        True           41
Name: count, dtype: int64

A única entrada incoerente realmente é a inclusão do *?* como resposta para o tipo de trabalho, e há uma quantidade bem significativa de entrada desse tipo, 2799 entradas, representando 5% do dataset.

In [31]:
setores = pd.Series(data['workclass'].unique())

for i in setores:
  print("Setor {}, quantidade de entradas = {}".format(i, data.workclass[data['workclass'] == i].count()))

Setor Private, quantidade de entradas = 33906
Setor Local-gov, quantidade de entradas = 3136
Setor nan, quantidade de entradas = 0
Setor Self-emp-not-inc, quantidade de entradas = 3862
Setor Federal-gov, quantidade de entradas = 1432
Setor State-gov, quantidade de entradas = 1981
Setor Self-emp-inc, quantidade de entradas = 1695
Setor Without-pay, quantidade de entradas = 21
Setor Never-worked, quantidade de entradas = 10


In [32]:
data.groupby('workclass')['workclass'].value_counts().sort_values()

workclass
Never-worked           10
Without-pay            21
Federal-gov          1432
Self-emp-inc         1695
State-gov            1981
Local-gov            3136
Self-emp-not-inc     3862
Private             33906
Name: count, dtype: int64